src/rhsmlib/cloud/_base_provider.py

```diff
@@ -280,20 +288,66 @@ def _get_metadata_from_cache(self) -> Union[str, None]:
         """
         raise NotImplementedError
 
-    def _get_data_from_server(self, data_type, url) -> Union[str, None]:
+    @staticmethod
+    def debug_print_http_request(request) -> None:
+        """
+        Print HTTP request that will be sent using requests Python package
+        :param request: prepared HTTP request
+        :return: None
+        """
+        yellow_col = '\033[93m'
+        blue_col = '\033[94m'
+        red_col = '\033[91m'
+        end_col = '\033[0m'
+        msg = blue_col + "Making request: " + end_col
+        msg += red_col + request.method + ' ' + request.url + end_col
+        if 'SUBMAN_DEBUG_PRINT_REQUEST_HEADER' in os.environ:
+            headers = ', '.join('{}: {}'.format(k, v) for k, v in request.headers.items())
+            msg += blue_col + " {{{headers}}}".format(headers=headers) + end_col
+        if 'SUBMAN_DEBUG_PRINT_REQUEST_BODY' in os.environ and request.body is not None:
+            msg += yellow_col + f" {request.body}" + end_col
+        print()
+        print(msg)
+        print()
+
+    @staticmethod
+    def debug_print_http_response(response) -> None:
+        """
+        Print HTTP request that was received using requests Python package
+        :param response: received HTTP response
+        :return: None
+        """
+        print('\n{code} {{{headers}}}\n{body}\n'.format(
+            code=response.status_code,
+            headers=', '.join('{key}: {value}'.format(key=k, value=v) for k, v in response.headers.items()),
+            body=response.text,
+        ))
+
+    def _get_data_from_server(self, data_type: str, url: str, headers: str = None) -> Union[str, None]:
         """
         Try to get some data from server using method GET
-        :data_type: string representing data type (metadata, signature, token)
+        :param data_type: string representing data type (metadata, signature, token, etc.)
         :param url: URL of the GET request
+        :param headers: optional headers parameters. When not set, then self.HTTP_HEADERS are used
         :return: String representing body, when status code is 200; Otherwise return None
         """
         log.debug(f'Trying to get {data_type} from {url}')
 
+        if headers is None:
+            headers = self.HTTP_HEADERS
+
+        http_req = requests.Request(method='GET', url=url, headers=headers)
+        prepared_http_req = self._session.prepare_request(http_req)
+        if 'SUBMAN_DEBUG_PRINT_REQUEST' in os.environ:
+            self.debug_print_http_request(prepared_http_req)
+
         try:
-            response = requests.get(url, headers=self.HTTP_HEADERS)
+            response = self._session.send(prepared_http_req, timeout=self.TIMEOUT)
         except requests.ConnectionError as err:
             log.debug(f'Unable to get {self.CLOUD_PROVIDER_ID} {data_type}: {err}')
         else:
+            if 'SUBMAN_DEBUG_PRINT_RESPONSE' in os.environ:
+                self.debug_print_http_response(response)
```

Comment:
Moving debug printing to own method is actually good idea. BTW: I don't see how to add hooks to `Session` globally. There is nothing `requests.Session.hooks` and I also don't think that it would be good idea.

Reply:
According to the documentation [1], it seems you can do:
```python
  self._session.hooks['response'].append(function)
```

[1] https://docs.python-requests.org/en/latest/user/advanced/#event-hooks

Reply:
Yeah, I do it in this way in the current version of code.